binary-decode.go

```diff
@@ -695,18 +696,158 @@ func (cdc *Codec) decodeReflectBinaryStruct(bz []byte, info *TypeInfo, rv reflec
 
 		// Read "StructTerm".
 		// NOTE: In the future, we'll need to break out of a loop
-		// when encoutering an StructTerm typ3 byte.
-		var typ = Typ3(0x00)
+		// when encountering an StructTerm typ3 byte.
+
+		var typ Typ3
+		// we have read all, or do we? consume the rest
+		for len(bz) > 2 && bz[0]&0xF8 != 0 && Typ3(bz[0]) != Typ3_StructTerm {
+			fnum, typ3, _n, _ := decodeFieldNumberAndTyp3(bz)
+			if slide(&bz, &n, _n) && err != nil {
+				return
+			}
+			if fnum > uint32(len(info.Fields)) {
```

Comment:
This conditional should rather be a sanity check, that fnum > `greatest field number seen`.  Instead of len(info.Fields) (which isn't forward-compatible once we support custom field tag numbers), above line 651 we should keep track of `var lastFieldNum` which we keep incrementing in both for loops (651 and 703), and we should return an error from either for-loop if each newly encountered field doesn't have a greater field number than `lastFieldNum`.

As is, this conditional block will just be ignored if fnum doesn't increment, but lets be more restrictive in our decoding.

Reply:
Perfect 💯Thanks a lot for the thorough review and the suggestions!